app/main.py

```diff
@@ -1,2 +1,108 @@
+from dataclasses import dataclass
+from typing import Hashable, Any, Optional
+from fractions import Fraction
+
+
 class Dictionary:
-    pass
+    _INITIAL_CAPACITY = 8
+    _RESIZE_THRESHOLD = Fraction(2, 3)
+    _CAPACITY_MULTIPLIER = 2
+
+    @dataclass
+    class Node:
+        key: Hashable
+        value: Any
+        hash_value: int
+
+    def __init__(
+            self,
+            capacity: int = _INITIAL_CAPACITY
+    ) -> None:
+        self.capacity = capacity
+        self.resize_threshold: Fraction = Dictionary._RESIZE_THRESHOLD
+        self.capacity_multyplier: int = Dictionary._CAPACITY_MULTIPLIER
+        self.size = 0
+        self.hash_table: list[Dictionary.Node | None] = [None] * self.capacity
+
+    def _calculate_index(self, key: Hashable, hash_value: int) -> int:
+        index = hash_value % self.capacity
+
+        while (
+                self.hash_table[index] is not None
+                and self.hash_table[index].key != key
+        ):
+            index = (index + 1) % self.capacity
+        return index
+
+    @property
+    def current_max_size(self) -> int:
+        return int(self.capacity * self.resize_threshold)
+
+    def resize(self) -> None:
+        old_hash_table = self.hash_table
+        self.capacity *= self.capacity_multyplier
+        self.size = 0
+        self.hash_table = [None] * self.capacity
+
+        for node in old_hash_table:
+            if node is not None:
+                self.__setitem__(node.key, node.value)
+
+    def __setitem__(self, key: Hashable, value: Any) -> None:
+        hash_value = hash(key)
+        index = self._calculate_index(key, hash_value)
+
+        if self.hash_table[index] is None:
+            if self.size + 1 >= self.current_max_size:
+                self.resize()
+                return self.__setitem__(key, value)
```

Comment:
refactor this so you don't have to make a recursive call. 
Check if table needs resizing first, resize if necessary, then calculate index and set node.

Reply:
I still think you can do it without recursion, you can even make it more simple with less if statements. Recursion can be dangerous and it makes your code harder to debug